### PR TITLE
some examples

### PR DESCRIPTION
**Pull request by KhrulkovV**
Started writing up examples

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,9 @@
 
 # Test coverage results.
 .coverage
+
+# Mac DS_Store files
+.DS_Store
```
Comment (Bihaqo) [inline .gitignore]: Please, remove DS_Store from here, it's usually not encouraged to include tmp files specific to your machine

File: examples/utils/tt_dense.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            h = t3f.matmul(x, self.W)
+        if self.activation is not None:
+            if self.activation in activations:
+                h = Activation(self.activation)(h)
+            else:
+                raise ValueError('unknown activation')
+        return h
+
+    def compute_output_shape(self, input_shape):
+        return (input_shape[0], self.output_dim)
```
Comment (Bihaqo) [inline examples/utils/tt_dense.py]: Also, why is it input_shape[0]? What is input_shape[1]?

File: t3f/initializers.py [hunk 3]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
...
+
+
+def random_matrix_xavier(shape, tt_rank=2):
+  """Constructs a random TT matrix with entrywise variance 2.0 / (n_in + n_out)
+
+  Args:
+    shape: 2d array, shape[0] is the shape of the matrix row-index,
+      shape[1] is the shape of the column index.
+      shape[0] and shape[1] should have the same number of elements (d)
+      Also supports ommiting one of the dimensions for vectors, e.g.
```
Comment (Bihaqo) [inline t3f/initializers.py]: omitting

File: t3f/initializers.py [hunk 4]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
...
+        random_matrix([[2, 2, 2], None])
```
Comment (Bihaqo) [inline t3f/initializers.py]: random_matrix_xavier

File: t3f/initializers.py [hunk 5]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
...
+      and
+        random_matrix_xavier([None, [2, 2, 2]])
+      will create an 8-element column and row vectors correspondingly.
+    tt_rank: a number.
+    mean: a number, the mean of the normal distribution used for
```
Comment (Bihaqo) [inline t3f/initializers.py]: Remove

File: t3f/initializers.py [hunk 6]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
...
+      initializing TT-cores.
+    stddev: a number, the standard deviation of the normal distribution used
```
Comment (Bihaqo) [inline t3f/initializers.py]: Remove

File: t3f/initializers.py [hunk 7]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
...
+
+  n_in = np.prod(shape[0])
+  n_out = np.prod(shape[1])
+  lamb = 2.0 / (n_in + n_out)
+
+  # Empirically entries of TT-Matrix have variance r^(d-1)) and mean 0.
+  # We scale each TT-core to obtain the desired Xavier variance
+  # Var_xavier = 2 / (n_in + n_out)
+
+  sigma = (lamb * tt_rank[1] ** (1.0 - num_dims)) ** (1.0 / (2 * num_dims))
```
Comment (Bihaqo) [inline t3f/initializers.py]: Why TT-rank is a vector but you use only one element? Please, check that the input tt_rank is a number and throw a NotImplementedError otherwise

File: t3f/initializers.py [hunk 8]
```diff
@@ -483,3 +483,79 @@ def zeros_like(tt):
       return matrix_zeros(shape)
     else:
       return tensor_zeros(shape[0, :])
+
+
+def random_matrix_xavier(shape, tt_rank=2):
```
Comment (Bihaqo) [inline t3f/initializers.py]: Please, test this function

File: examples/tt-basic.ipynb [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+   "execution_count": 2,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "d = 10\n",
+    "x = np.linspace(0, 2 * np.pi, 2 ** 10)\n",
+    "y = np.sin(x)\n",
+    "y = np.reshape(y, d * [2], order='f')"
```
Comment (Bihaqo) [inline examples/tt-basic.ipynb]: This is scary. Why Fortran order?

File: examples/tt-basic.ipynb [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+   ],
+   "source": [
+    "print (y.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now lets construct a TT tensor out of `y` using `t3f.decompositions.to_tt_tensor`."
```
Comment (Bihaqo) [inline examples/tt-basic.ipynb]: t3f.to_tt_tensor instead of t3f.decompositions.to_tt_tensor

File: examples/tt-basic.ipynb [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "relative error is 2.9627586451526042e-15\n"
+     ]
+    }
+   ],
+   "source": [
+    "with tf.Session() as sess:\n",
```
Comment (Bihaqo) [inline examples/tt-basic.ipynb]: I think it is beater to use one session for the whole notebook (maybe interactive session?), since starting a session is slow, especially with a GPU

File: examples/tt-basic.ipynb [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    "    y_tt_round_full = sess.run(t3f.full(y_tt_round))\n",
+    "    print (\"relative error is {}\".format(np.linalg.norm(y_tt_round_full - y) / np.linalg.norm(y)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This fact allows us to significantly compress the tensor `y`. The total number of degrees of freedom in `y_tt` is :"
```
Comment (Bihaqo) [inline examples/tt-basic.ipynb]: in y_tt after rounding

File: examples/tensor-nets.ipynb [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "A TT-Matrix of size 784 x 625, underlying tensor shape: (4, 7, 4, 7) x (5, 5, 5, 5), TT-ranks: (1, 2, 2, 2, 1)\n"
+     ]
+    }
+   ],
+   "source": [
+    "W = t3f.random_matrix([[4, 7, 4, 7], [5, 5, 5, 5]], tt_rank=2)\n",
```
Comment (Bihaqo) [inline examples/tensor-nets.ipynb]: xavier

File: examples/tensor-nets.ipynb [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "x = tf.placeholder(tf.float32, [None, 784])\n",
+    "y = tf.placeholder(tf.int64, [None])\n",
+    "\n",
+    "W1 = t3f.get_variable('W1', initializer=W) \n",
+    "b1 = tf.get_variable('b1', shape=[625])\n",
+    "h1 = t3f.matmul(x, W1) + b1\n",
```
Comment (Bihaqo) [inline examples/tensor-nets.ipynb]: Add ReLU

File: examples/utils/tt_dense.py (hunk 2 above)
Comment (KhrulkovV) [reply examples/utils/tt_dense.py]: input_shape corresponds to the shape of this layer's input tensor (so input_shape[0] is typically the batch size), this function is used by keras to compute shapes automatically.
In general [input_shape[0], input_shape[1] , input_shape[2]...] -> [input_shape[0], self.output_dim]

File: t3f/tensor_train_base.py [hunk 15]
```diff
@@ -151,6 +151,28 @@ def __add__(self, other):
     from t3f import ops
     return ops.add(self, other)
 
+  def __sub__(self, other):
```
Comment (Bihaqo) [inline t3f/tensor_train_base.py]: Its a bad practice to pack lots of unrelated stuff in one pull request. Its ok for now, but please avoid in the future (e.g. you could split this in 3 pull requests)

File: examples/tensor-completion.ipynb [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+    "# but we would rather generate indecies of observed cells.\n",
+    "\n",
+    "ratio = 0.001\n",
+    "\n",
+    "# Let us simply randomly pick some indecies (it may happen\n",
+    "# that we will get duplicates but probability of that\n",
+    "# is 10^(-14) so lets not bother for now)\n",
+    "\n",
+    "num_observed = int(ratio * total_observed)\n",
+    "observation_idx = np.random.randint(0, 10, size=(num_observed, len(shape)))\n",
```
Comment (Bihaqo) [inline examples/tensor-completion.ipynb]: np.random.seed(0)

File: examples/tensor-riemannian.ipynb [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# Initialize A randomly, with large tt-ranks\n",
+    "shape = 10 * [2]\n",
```
Comment (Bihaqo) [inline examples/tensor-riemannian.ipynb]: Fix random seed

File: t3f/initializers.py [hunk 18]
```diff
@@ -483,3 +485,465 @@ def zeros_like(tt):
...
+      raise ValueError('all elements in `tt_rank` should be integers')
+
+  n_in = np.prod(shape[0])
+  n_out = np.prod(shape[1])
+  lamb = 2.0 / (n_in + n_out)
+
+  return random_matrix(shape, tt_rank=tt_rank, stddev=np.sqrt(lamb))
+
+
+def he(shape, tt_rank=2):
```
[the remark this answers is not in the document]
Comment (Bihaqo) [reply t3f/initializers.py]: I mean its short and not descriptive

**Comment by Bihaqo** [comment]
Also, would be amazing to add to the tensor net example how to train a normal NN, save to checkpoint, load from checkopoint, substitute one layer with TT-layer (initializing with t3f.to_tt_tensor(trained_W)), and finetuning.

File: t3f/initializers_test.py [hunk 19]
```diff
@@ -32,7 +33,7 @@ def testMatrixOnesAndZeros(self):
     zeros_desired = np.zeros((24, 10))
 
     bad_shapes = [[[-1, 2, 3], [3, 4, 6]], [[1.5, 2, 4], [2, 5, 6]],
-                  [[1], [2, 3]]]
+                  [[1], [2, 3]], [2, 3, 4]]
```
[the remark this answers is not in the document]
Comment (KhrulkovV) [reply t3f/initializers_test.py]: To cover more exceptions (so coveralls can be happier)

File: t3f/initializers_test.py [hunk 20]
```diff
@@ -18,10 +18,11 @@ def testTensorOnesAndZeros(self):
       tt_zeros_full = sess.run(ops.full(tt_zeros))
       self.assertAllClose(tt_ones_full, ones_desired)
       self.assertAllClose(tt_zeros_full, zeros_desired)
-    bad_shapes = [[[2, 3]], [-1.0, 3]]
+    bad_shapes = [[[2, 3]], [-1, 3], [0.1, 4]]
```
[the remark this answers is not in the document]
Comment (KhrulkovV) [reply t3f/initializers_test.py]: To cover more exceptions (so coveralls can be happier). Negative input + fractional input separately

File: t3f/initializers.py [hunk 21]
```diff
@@ -483,3 +485,465 @@ def zeros_like(tt):
...
+
+  Returns:
+    TensorTrain containing a TT-matrix of size
+      np.prod(shape[0]) x np.prod(shape[1])
+  """
+
+  # In case the shape is immutable.
+  shape = list(shape)
+  # In case shape represents a vector, e.g. [None, [2, 2, 2]]
+  if shape[0] is None:
```
[the remark this answers is not in the document]
Comment (KhrulkovV) [reply t3f/initializers.py]: Fixed

File: t3f/initializers.py [hunk 22]
```diff
@@ -7,6 +7,53 @@
 from t3f import shapes
 
 
+def _validate_input_parameters(**params):
```
Comment (Bihaqo) [inline t3f/initializers.py]: Make is_tensor and shape required params:
_validate_input_parameters(is_tensor, shape, **params):

File: t3f/initializers.py [hunk 23]
```diff
@@ -7,6 +7,53 @@
...
+  """Internal function for validating input parameters
+  """
```
Comment (Bihaqo) [inline t3f/initializers.py]: Add description of parameters, epsessially "is_tensor" since it's confusing

File: t3f/initializers.py [hunk 24]
```diff
@@ -7,6 +7,53 @@
...
+        raise ValueError('`tt_rank` should be integer')
+    if tt_rank.size > 1:
+      if not all(isinstance(tt_r, np.integer) for tt_r in tt_rank):
+        raise ValueError('all elements in `tt_rank` should be integers')
+    if np.any(tt_rank < 1):
+      raise ValueError('`rank` should be positive')
+
+    if params['is_tensor']:
+      if tt_rank.size != 1 and tt_rank.size != (shape.size + 1):
+        raise ValueError('`rank` array has inappropriate size')
```
Comment (Bihaqo) [inline t3f/initializers.py]: `tt_rank` array has inappropriate size: expected 1 or %d, got %d

File: t3f/initializers.py [hunk 25]
```diff
@@ -7,6 +7,53 @@
...
+    else:
+      if tt_rank.size != 1 and tt_rank.size != (shape[0].size + 1):
+        raise ValueError('`rank` array has inappropriate size')
```
Comment (Bihaqo) [inline t3f/initializers.py]: `tt_rank` array has inappropriate size: expected 1 or %d, got %d

File: t3f/initializers.py [hunk 26]
```diff
@@ -7,6 +7,53 @@
...
+  if 'tt_rank' in params:
+    tt_rank = params['tt_rank']
+    if tt_rank.size == 1:
+      if not isinstance(tt_rank[()], np.integer):
+        raise ValueError('`tt_rank` should be integer')
+    if tt_rank.size > 1:
+      if not all(isinstance(tt_r, np.integer) for tt_r in tt_rank):
+        raise ValueError('all elements in `tt_rank` should be integers')
+    if np.any(tt_rank < 1):
+      raise ValueError('`rank` should be positive')
```
Comment (Bihaqo) [inline t3f/initializers.py]: tt_rank should be positive, got %a

File: t3f/initializers.py [hunk 27]
```diff
@@ -7,6 +7,53 @@
...
+    if batch_size < 1:
+      raise ValueError('Batch size should be positive')
+  if 'tt_rank' in params:
+    tt_rank = params['tt_rank']
+    if tt_rank.size == 1:
+      if not isinstance(tt_rank[()], np.integer):
+        raise ValueError('`tt_rank` should be integer')
+    if tt_rank.size > 1:
+      if not all(isinstance(tt_r, np.integer) for tt_r in tt_rank):
+        raise ValueError('all elements in `tt_rank` should be integers')
```
Comment (Bihaqo) [inline t3f/initializers.py]: , got %a

File: t3f/initializers.py [hunk 28]
```diff
@@ -7,6 +7,53 @@
...
+    batch_size = params['batch_size']
+    if not isinstance(batch_size, (int, np.integer)):
+      raise ValueError('`batch_size` should be integer')
+    if batch_size < 1:
+      raise ValueError('Batch size should be positive')
+  if 'tt_rank' in params:
+    tt_rank = params['tt_rank']
+    if tt_rank.size == 1:
+      if not isinstance(tt_rank[()], np.integer):
+        raise ValueError('`tt_rank` should be integer')
```
Comment (Bihaqo) [inline t3f/initializers.py]: , got %d

File: t3f/initializers.py [hunk 29]
```diff
@@ -7,6 +7,53 @@
...
+      raise ValueError('all elements in `shape` should be positive')
+    if not all(isinstance(sh, np.integer) for sh in shape.flatten()):
+      raise ValueError('all elements in `shape` should be integers')
+
+  if 'batch_size' in params:
+    batch_size = params['batch_size']
+    if not isinstance(batch_size, (int, np.integer)):
+      raise ValueError('`batch_size` should be integer')
+    if batch_size < 1:
+      raise ValueError('Batch size should be positive')
```
Comment (Bihaqo) [inline t3f/initializers.py]: , got %d

File: t3f/initializers.py [hunk 30]
```diff
@@ -7,6 +7,53 @@
...
+      raise ValueError('shape[0] should have the same length as shape[1]')
+    if np.any(shape.flatten() < 1):
+      raise ValueError('all elements in `shape` should be positive')
+    if not all(isinstance(sh, np.integer) for sh in shape.flatten()):
+      raise ValueError('all elements in `shape` should be integers')
+
+  if 'batch_size' in params:
+    batch_size = params['batch_size']
+    if not isinstance(batch_size, (int, np.integer)):
+      raise ValueError('`batch_size` should be integer')
```
Comment (Bihaqo) [inline t3f/initializers.py]: , got %d

File: t3f/initializers.py [hunk 31]
```diff
@@ -7,6 +7,53 @@
...
+      raise ValueError('all elements in `shape` should be integers')
+  else:
+    if len(shape.shape) != 2:
+      raise ValueError('shape should be 2d array')
+    if shape[0].size != shape[1].size:
+      raise ValueError('shape[0] should have the same length as shape[1]')
+    if np.any(shape.flatten() < 1):
+      raise ValueError('all elements in `shape` should be positive')
+    if not all(isinstance(sh, np.integer) for sh in shape.flatten()):
+      raise ValueError('all elements in `shape` should be integers')
```
Comment (Bihaqo) [inline t3f/initializers.py]: Here as well (see below)
